### DISCUSSION TITLE
Filling face-centred boundary data

### DISCUSSION THREAD

**Discussion by ankithadas** [AMReX-Codes/amrex#3561] [Q&A] [answered]
Hey AMReX team,

I am trying to write up a steady state Navier Stokes solver using AMReX. I am trying to implement something similar to SIMPLE scheme but with AMR support. I understand that cell centered boundary conditions can be filled with PhysBCFunct<GpuBndryFuncFab<obj>> which fills the values at the ghost nodes around the domain. How can I fill physical boundary data on face centered data (advection velocity) ? Any suggestions would be much appreciated! 

Kind regards,
Ankith

**Comment by drummerdoc** [comment]
BCs are a bit nuanced in this context.  A user loads Dirichlet physical boundary conditions to be applied ON THE FACES, but puts that data into a cell-centered container, primarily due to coding convenience.  That is, if I want to set velocity at inflow, I will put the values into a cell-centered grow cell, but the code internally will apply that at the face.  That said, much of this implementation is in code that some might consider user code (such as incflo, for example, or PeleLMeX, or IAMR).  The library itself is typically considered to be the data structures, parallel communication stuff, looping constructs and IO.  It's a bit gray and fuzzy.  I mention this because you are actually writing one of these applications as your own solver.  Depending on which pieces you pull in to your application, some of the data centering decisions might be made by you in your stencil codes.  If you use the AMReX-Hydro routines for advection and the MLMG stuff for diffusion operators for example however, you'll be able to see where those decision have been made for you. (e.g., applyBC in the linear operators for diffusion).  Let me know if this need more explanation.  Lots to unpack here.

**Reply by ankithadas** [reply]
Thank you for your detailed explanation. To give some context, I plan to use MLMG for both diffusion and advection (custom implementation) in NS solver. For filling ghost cells for cell-centered data, I am currently using a struct with `void operator()` implemented. It looks something like this 
```c++
void operator() (const amrex::IntVect& iv, amrex::Array4<amrex::Real> const& vel,
                 const int dcomp, const int num_comp,
                 amrex::GeometryData const& geom, const amrex::Real /*time*/,
                 const amrex::BCRec* bcr, const int bcomp,
                 const int orig_comp) const
{
    const auto& [i, j, k] = iv.toArray();  
    const auto& domain = geom.Domain();    
    for (int icomp = 0; icomp < num_comp; ++icomp ) {
        const auto& bc = bcr[bcomp+icomp];
        if (bc.lo(0) == amrex::BCType::ext_dir && i < domain.smallEnd(0)) {
            vel(i,j,k,icomp) = 4; // Fill in the ghost cell with 4 
        }
    }
}
```
With this technique of filling boundary cells, I would not need to change the stencil for cell-centered data and I just need to keep the boundaries updated. It is my understanding that I can manually fill the domain boundary using appropriate boxes around the boundary. However, I wanted to know if there was a similar mechanism for face-centered data already implement in amrex framework. 

Again, thank you for all the help in advance!

Kind regards,
Ankith

**Reply by drummerdoc** [reply, marked as answer]
for the MLMG framework, there are some "expert" settings that set where (in normal distance from the bounding face)  the BC is to be applied.  A lot of that is done automatically behind the scenes but can be over-ridden for specific applications.  Things have moved around in the code a little since I worked in that area, but it may be worth digging through the examples a little to find where this is done (used to be called bcloc, and was a scalar for each orientation).  So, the Dirichlet values were passed in the cell-centered container, and bcloc was used to decide *where* the value was to be applied.  It may be that the defaults do exactly what you want, but it's good to be sure.
